base/util/man/man1/store_dir.1

.\"
.TH store_dir 1  26-November-2007  "DRMS MANPAGE" "DRMS Programmer's Manual"
.SH NAME
store_dir, retrieve_dir, store_file, retrieve_file \- store files in and retrieve files from a SUMS directory.
.SH SYNOPSIS
.nf
.B store_dir \fR[\fB-cv\fRDRIVER_FLAGS] \fBseries=\fR<series_name> \fBdirname=\fR<prefix>/<subdir> \fBsel=\fR<sel_text> [\fBnote=\fR<note_text>] [\fBperm=0\fR|\fB1\fR]
.B retrieve_dir \fR[\fB-l\fRDRIVER_FLAGS] \fBseries=\fR<record_set> \fBto=\fR<dest> [\fBdirkey=\fR<dirname_keyword>]
.B store_file \fR[\fB-chv\fRDRIVER_FLAGS] \fBds=\fR<series_name> \fBin=\fR<file> [\fBsel=\fR<sel_text>] [\fBnote=\fR<note_text>] [\fBperm=0\fR|\fB1\fR]
.B retrieve_file \fR[\fB-fhlv\fRDRIVER_FLAGS] \fBds=\fR<record_set> \fBout=\fR<dest> [\fBsegment=\fR<segment_name>]

.SH DESCRIPTION
.PP
store_dir and retrieve_dir are a pair of utilities to store (and retrieve) an arbitrary 
directory of files to (from) DRMS/SUMS. Files stored with store_dir can be easily retrieved by 
the retrieve_dir utility. store_file and retrieve_file provide the analogous functionality
for a single file.
.PP
If the series <series_name> does not exist, store_dir will first create that series,
provided that the user supplies the \fBc\fR flag. If \fBperm=1\fR, then the series will be 
globally accessible, otherwise, only the user calling store_dir will have access.
store_dir then creates a record in the series <series_name> and creates a subdirectory, 
<subdir>, in the record's SUMS directory (for each record there is a single SUMS 
directory). store_dir copies all files from <prefix>/<subdir> to <subdir> in the 
record's SUMS directory. It also stores <prefix>/<subdir> 
in the record's \fBdirname\fR keyword value so that retrieve_dir can locate the subdirectory
within the SUMS directory that contains the files. sel=<sel_text> and note=<note_text> 
allow the user to link additional information to the record to facilitate the subsequent
search for files in the series. They set the record's \fBsel\fR and \fBnote\fR keywords 
to have the values <sel_text> and <note_text>, respectively. In particular, \fB<sel>\fR
can be used to differentiate between multiple calls to store_dir with the same value 
for <prefix>/<subdir>.
The series <series_name> 
must contain prime keywords named \fBdirname\fR and \fBsel\fR, and it must contain the keyword
named \fBnote\fR.
.PP
For retrieve_dir, <record_set> is a database query that allows the user to select 
a subset of records of a series. In particular, the user can supply values for the
\fBsel\fR and \fBnote\fR keywords in <record_set> to search for files saved with store_dir.
retrieve_dir retrieves the set of records specified, 
and for each record uses the value contained within the <dirname_keyword> keyword 
(defaults to \fBdirname\fR) to
identify the subdirectory within the record's SUMS directory (i.e., 
it is all of the string after the last '/' in this value) that contains the record's 
files. It then copies (or links, if the \fBl\fR flag is present) all files from
this subdirectory to a identically named subdirectory of the directory specified by <dest>.
Typically, the files retrieved with retrieve_dir have been stored into SUMS
with the store_dir utility.  retrieve_dir will overwrite any files existing in the destination 
subdirectory whose name matches the name of a file to be copied from
a SUMS subdirectory.  If the link flag, \fBl\fR,
is specified, then retrieve_dir will create a link to the SUMS subdirectory instead of 
copying from the SUMS subdirectory to the destination subdirectory.  Keep in mind that over time
this link may become broken as SUMS may remove a stored file if it is present in SUMS 
longer than its specified retention time.
.PP
For store_file, <file> is the full path of the file that will be stored in SUMS
in the dataseries <series_name>. The series <series_name> must contain prime keywords
named \fBfile\fR and \fBsel\fR, and it must contain the keyword named \fBnote\fR.
.PP
For each file in the record set specified by <record_set>, retrieve dir will copy the
record segment's file to <dest>. If the series contains more than one segment, <segment_name>
must be present.  Since the name of a segment's file is the segment name,
any file in <dest> whose name matches the segment's name will be overwritten. However,
a backup will be saved (the file will be renamed with a \fB~\fR suffix) if an overwrite it 
to occur, unless the \fBf\fR flag is specified.  retrieve_dir prints to stdout the
list of files retrieved, or NOT_FOUND if a file cannot be found. It also prints the contents
of the \fBnote\fR keyword value.
.PP
Example to store a directory of files:

  store_dir series=su_arta.TestStoreDir dirname=/auto/home2/arta/savedFilesJanuary sel=January 
note="All my January work"

Example to retrieve a directory of files:

  retrieve_dir series=su_arta.TestStoreDir to=/auto/home2/arta/restoredFilesJanuary/ dirkey=dirname

Example to store a single file:

  store_file in=0000.fits ds=myseries sel=test002 note= "fits file from test002"

Example to retrieve a single file:

  retrieve_file ds=myseries[][test002]

.SH FLAGS
.SS module-specific flags
.IP \fB-c\fR 20
Silently create the series <series_name> if it does not exist.
.IP \fB-f\fR 20
Force removal of a pre-existing file with same name as retreived file.
.IP \fB-h\fR 20
Print usage message and exit
.IP \fB-l\fR 20
Create a symbolic link to the SUMS directory containing the file(s) of interest.
.IP \fB-v\fR 20
Run in verbose mode.
.SS driver flags
.IP \fB-H\fR|\fB--help\fR 20
Show jsoc_main usage information.
.IP \fB-L\fR 20
Run jsoc_main driver without logging.
.IP \fB-V\fR 20
Run jsoc_main driver in verbose mode.
.IP \fB-Q\fR 20
Run jsoc_main driver in quiet mode (no terminal output).

.SH ARGUMENTS
.IP <series_name> 20 
Specifies the series, <series_name>, to which the file-referring record will be added.
Should a record-set specifier be provided, it will be ignored.
.IP <prefix>/<subdir> 20
The local full path that contains the files to be stored in SUMS. The record created will contain 
a prime keyword named \fBdirname\fR whose value will be <prefix>/<subdir>.
.IP <sel_text> 20
<sel_text> contains a string that will be the keyword value for the \fBsel\fR prime keyword of the 
record created. This extra prime keyword facilitates selection between multiple 
records containing equivalent values for the \fBdirname\fR (store_dir) or /fBfile\fR (store_file)
keyword (it may be desirable to save files in the same directly repeatedly, or save the same
file over time).
.IP <note_text> 20
An optional string that will be the keyword value for the \fBnote\fR keyword of the record created.
.IP \fBperm=0\fR|\fB1 20
Relevant only when the \fBc\fR flag is used and a series is created.  A perm of 0 will make the 
created series accessible only by the current user.  A perm of 1 will make the series globally
accessible.
.IP <record_set> 20
A series name followed by an optional record set filter 
(i.e., <name>[RecordSet_filter]). Causes selection of a subset of 
records in the series. Each record retrieved with this <record_set> query
will refer to one directory/file that will be copied to <dest>.
.IP <dest> 20
The destination directory to which the SUMS file(s) should be copied/linked.  
Do not append a final '/' to the path (i.e., /home/arta/dir1 is acceptable, but /home/arta/dir1/
is not)
.IP <dirname_keyword> 20
Name of the keyword containing the path of the directory originally
stored with store_dir.
.IP <file> 20
Full path of the file to store. Only the filename (everything after the last '/') 
will be stored in the \fBfile\fR keyword.
.IP <segment_name> 20
Name of the segment whose file should be retrieved.


src/base/util/apps/store_dir.c
/* store_dir - stores a dir in DRMS/SUMS. */

/* Usage:
 *  store_dir series=<name> dirname=<dir> sel=<selection key> {note=<comment>}
 *         
 *     Store a dir into SUMS.
 *     "dir" is the directory that will be copied in SUMS
 *     in the dataseries <name>.  
 *     No <record spec> (i.e. no "[...]") should be given with the series.
 *     If <note> is present the comment will be stored with the data.
 *     The keyword "sel" can be used to provide lookup selection info in case
 *     multiple dirs with the same series will be stored.
 *     The DRMS prime keys will the values for "dir" and "sel".  
 *     Multiple calls of
 *     store_dir with the same seriesname, dir, and sel will result in
 *     multiple records with the same prime key.  Only the most recent
 *     will be easily retrieved. 
 *     If the dir in the "dir" keyword is sufficient it can be used as
 *     the prime key.  
 *
 *   store_dir series=su_jim.storedirs dirname=/home/xx sel=test1 note="junk"
 *
 *     will be retrieved with simply:
 *
 *       retrieve_dir su_jim.storedirs[/home/xx] to=/tmp
 *  or   retrieve_dir su_jim.storedirs[][test1] to=/tmp
 *
 *    Prior to first use, the series should be created. 
 *    String keywords of "dirname", "sel", and "note" should be included.
 *    If the series does not exist it is made from a standard template if the 
 *    -c flag is given or if the user answers "yes" when asked.  
 *    If a series is made the parameter "perm" supplies the series access 
 *    permissions.  Default is 1 which is like "perm=s" for create_series.  
 *    Set perm to 0 if you want to exclude other users from
 *    accessing your stored dir.
 */

#include "jsoc_main.h"
#include "drms.h"
#include "drms_names.h"
#include <unistd.h>
#include <strings.h>

#define NOTSPECIFIED "***NOTSPECIFIED***"
/* List of default parameter values. */
ModuleArgs_t module_args[] = { 
  {ARG_STRING, "series", "", "Series name to store the dir into"},
  {ARG_STRING, "dirname", "", "Dir to store. prime key"},
  {ARG_STRING, "sel", "", "selection name. prime key"}, 
  {ARG_STRING, "note", "N/A", "comment field"},
  {ARG_INT, "perm", "1", ""}, /* permissions in case series is created */
  {ARG_FLAG, "c", "0", "create new series if needed"},
  {ARG_FLAG, "v", "0", "verbose flag"},
  {ARG_END}
};

/* Module name presented to DRMS. */
char *module_name = "store_dir";

/* Some global variables for this module. */
int verbose = 0;

/* Check DRMS session status and calling arguments and set verbose variable */
int nice_intro () {
  verbose = cmdparams_get_int (&cmdparams, "v", NULL);
/********************this is old stuff before Rick's arg handling *********
  int usage = cmdparams_get_int (&cmdparams, "h", NULL);
  if (usage) {
    printf ("store_dir <series> {-c} {-h} {-v} in=<dir> {sel=<desc>} {note=<comment>} {perm=1}\n"
	"  -c: Quietly create series if does not exist\n"
	"  -h: print this message and exit\n"
	"  -v: verbose\n"
	"series - series to store dir into\n"
	"in=    - dir to store\n"
	"sel=   - optional description for retrieval key\n"
	"note=  - optional description for dir\n"
        "perm=1 - optional access permissions, use 0 for private.\n");
    return (1);
  }
**********************************************************************/
  if (verbose) cmdparams_printall (&cmdparams);
  return(0);
}

/* Module main function. */
int DoIt (void) {
int status = 0;

char *in, *series, *note, *dirname, *sel, *rsp;
char path[DRMS_MAXPATHLEN];
char cmd[DRMS_MAXPATHLEN+1024];
DRMS_Record_t *rec, *template;
int yes_create = cmdparams_get_int(&cmdparams, "c", NULL);

if (nice_intro())
  return(0);

/*series = strdup(cmdparams_getarg(&cmdparams, 1));*/
series = cmdparams_get_str(&cmdparams, "series", NULL);
note = cmdparams_get_str(&cmdparams, "note", NULL);
sel = cmdparams_get_str(&cmdparams, "sel", NULL);
in = cmdparams_get_str(&cmdparams, "dirname", NULL);
/* First, check to see that the dir exists */
if (access(in, R_OK) != 0)
  {
  printf("The requested dir can not be accessed.  %s\n", in);
  return(1);
  }
  dirname = in;

/* Look in the record spec to see if a sel is given, else use dirname */
rsp = index(series, '[');
if (rsp)
  {
  /* a "[" was present, ignore it */
  *rsp++ = '\0';
  printf("WARNING: Destination series %s was given a record spec: %s, which will be ignored.\n",series,rsp);
  }

/* Check to see if series exists, make if should */
template = drms_template_record(drms_env, series, &status);
if (template==NULL && status == DRMS_ERROR_UNKNOWNSERIES)
  {
  int user_says_yes();
  if (yes_create || user_says_yes(series))
    {
    int create_series(char *series);
    if (create_series(series))
      {
      printf("Series '%s' does not exist. Create_series failed. Give up.\n", series);
      return(1);
      }
    }
  else
    {
    printf("Series '%s' does not exist. Give up.\n", series);
    return(1);
    }
  }
else
  if(status)
    {
    fprintf(stderr, "DRMS problem looking up series.\n");
    return(status);
    }

/* Now ready to make a new record and set keywords */
rec = drms_create_record(drms_env, series, DRMS_PERMANENT, &status);
if (!rec || status)
    {
    printf("drms_create_record failed, series=%s, status=%d.  Aborting.\n", series,status);
    return(status);
    }
if ((status = drms_setkey_string(rec, "dirname", dirname)))
   {
   if (status == DRMS_ERROR_UNKNOWNKEYWORD)
     printf("ERROR: series %s does not have a keyword named 'dirname'\n", series);
   else
     printf("ERROR: drms_setkey_string failed for 'dirname'\n");
   return(1);
   }
if ((status = drms_setkey_string(rec, "sel", sel)))
   {
   if (status == DRMS_ERROR_UNKNOWNKEYWORD)
     printf("ERROR: series %s does not have a keyword named 'sel'\n", series);
   else
     printf("ERROR: drms_setkey_string failed for 'sel'\n");
   return(1);
   }
if ((status = drms_setkey_string(rec, "note", note))) 
   {
   if (status == DRMS_ERROR_UNKNOWNKEYWORD)
     printf("WARNING: series %s does not have a keyword named 'note'\n", series);
   else
     printf("WARNING: drms_setkey_string failed for 'note'\n");
   printf("Your note was not saved: %s\n",note);
   }
/* a SU should have been allocated since there is a segment in the series defn */
drms_record_directory(rec, path, 1);
if (! *path)
  {
  fprintf(stderr,"no path found\n");
  return(1);
  }
sprintf(cmd, "cp -rp %s %s", in, path);
status = system(cmd);
if (status)
    {
    printf("Copy failed: %s\n", cmd);
    return(status);
    }
else
    printf("Command success: %s\n", cmd);
if ((status = drms_close_record(rec, DRMS_INSERT_RECORD)))
  printf("drms_close_record failed!\n");

return(status);
}

int user_says_yes(char *series)
  {
  char resp[DRMS_MAXPATHLEN];
  printf("Series '%s' does not exist. Do you want a standard store_dir series to be created?\n", series);
  printf("Answer yes or no: ");
  fflush(stdout);
  gets(resp);
  return(strcasecmp(resp,"yes")==0);
  }

char * make_series_jsd(char *series)
  {
  char *user = getenv("USER");
  char jsd[1024];
  sprintf(jsd,
    "Seriesname:     %s\n"
    "Author:         %s\n"
    "Owners:         %s\n"
    "Unitsize:       1\n"
    "Archive:        1\n"
    "Retention:      7\n"
    "Tapegroup:      1\n"
    "Index:          dirname,sel\n"
    "Description:    \"Dir storage\"\n"
    "Keyword: dirname,  string, variable, record, \" \",  %%s, none, \"Original dirname\"\n"
    "Keyword: sel, string, variable, record, \" \",  %%s, none, \" \"\n"
    "Keyword: note,      string, variable, record, \" \",  %%s, none, \" \"\n"
    "Data: dir, constant, int, 0, none, generic, \" \"\n",
    series, user, user);
  return (strdup(jsd));
  }

int create_series(char *series)
  {
  char *jsd;
  int perm = cmdparams_get_int(&cmdparams, "perm", NULL);
  DRMS_Record_t *template;
  jsd = make_series_jsd(series);
  template = drms_parse_description(drms_env, jsd);
  if (template==NULL)
    {
    fprintf(stderr, "Failed to parse series description.  JSD was:\n%s\n",jsd);
    return(1);
    }
  if (drms_create_series(template, perm))
    {
    fprintf(stderr, "Failed to create series. JSD was:\n%s\n", jsd);
    return(1);
    }
  if (verbose)
    printf("Create_series successful, JSD was:\n%s\n", jsd);
  drms_free_record_struct(template);
  free(template);
  free(jsd);
  return(0);
  }
  
